File: Algorithm.CSharp/OptionAssignmentRegressionAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

using System.Collections.Generic;
using System.Linq;
using QuantConnect.Data;
using QuantConnect.Interfaces;
using QuantConnect.Securities;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// This regression algorithm verifies automatic option contract assignment behavior.
    /// </summary>
    /// <meta name="tag" content="regression test" />
    /// <meta name="tag" content="options" />
    /// <meta name="tag" content="using data" />
    /// <meta name="tag" content="filter selection" />
    public class OptionAssignmentRegressionAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
    {
        private Security Stock;

        private Security CallOption;
        private Symbol CallOptionSymbol;

        private Security PutOption;
        private Symbol PutOptionSymbol;

        public override void Initialize()
        {
            SetStartDate(2015, 12, 23);
            SetEndDate(2015, 12, 28);
            SetCash(100000);
            Stock = AddEquity("GOOG", Resolution.Minute);

            var contracts = OptionChainProvider.GetOptionContractList(Stock.Symbol, UtcTime).ToList();

            PutOptionSymbol = contracts
                .Where(c => c.ID.OptionRight == OptionRight.Put)
                .OrderBy(c => c.ID.Date)
                .First(c => c.ID.StrikePrice == 800m);

            CallOptionSymbol = contracts
                .Where(c => c.ID.OptionRight == OptionRight.Call)
                .OrderBy(c => c.ID.Date)
                .First(c => c.ID.StrikePrice == 600m);

            PutOption = AddOptionContract(PutOptionSymbol);
            CallOption = AddOptionContract(CallOptionSymbol);
        }

        public override void OnData(Slice data)
        {
            if (!Portfolio.Invested && Stock.Price != 0 && PutOption.Price != 0 && CallOption.Price != 0)
            {
                // this gets executed on start and after each auto-assignment, finally ending with expiration assignment
                if (Time < PutOptionSymbol.ID.Date)
                {
                    MarketOrder(PutOptionSymbol, -1);
                }

                if (Time < CallOptionSymbol.ID.Date)
                {
                    MarketOrder(CallOptionSymbol, -1);
                }
            }
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public bool CanRunLocally { get; } = true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public virtual Language[] Languages { get; } = { Language.CSharp };

        /// <summary>
        /// Data Points count of all timeslices of algorithm
        /// </summary>
        public long DataPoints => 4025;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public virtual Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Orders", "32"},
            {"Average Win", "9.60%"},
            {"Average Loss", "-16.91%"},
            {"Compounding Annual Return", "-84.836%"},
            {"Drawdown", "2.900%"},
            {"Expectancy", "-0.608"},
            {"Start Equity", "100000"},
            {"End Equity", "97114"},
            {"Net Profit", "-2.886%"},
            {"Sharpe Ratio", "-7.473"},
            {"Sortino Ratio", "0"},
            {"Probabilistic Sharpe Ratio", "1.125%"},
            {"Loss Rate", "75%"},
            {"Win Rate", "25%"},
            {"Profit-Loss Ratio", "0.57"},
            {"Alpha", "-0.016"},
            {"Beta", "0.458"},
            {"Annual Standard Deviation", "0.014"},
            {"Annual Variance", "0"},
            {"Information Ratio", "5.991"},
            {"Tracking Error", "0.015"},
            {"Treynor Ratio", "-0.229"},
            {"Total Fees", "$16.00"},
            {"Estimated Strategy Capacity", "$710000.00"},
            {"Lowest Capacity Asset", "GOOCV 305RBQ20WHPNQ|GOOCV VP83T1ZUHROL"},
            {"Portfolio Turnover", "218.80%"},
            {"OrderListHash", "84dbd164000f9fb9096c14cbcd0b3e15"}
        };
    }
}

File: Algorithm.Python/OptionAssignmentRegressionAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

### <summary>
### This regression algorithm verifies automatic option contract assignment behavior.
### </summary>
class OptionAssignmentRegressionAlgorithm(QCAlgorithm):

    def initialize(self):
        self.set_start_date(2015, 12, 23)
        self.set_end_date(2015, 12, 28)
        self.set_cash(100000)
        self.stock = self.add_equity("GOOG", Resolution.MINUTE)

        contracts = list(self.option_chain(self.stock.symbol))

        self.put_option_symbol = sorted(
            [c for c in contracts if c.id.option_right == OptionRight.PUT and c.id.strike_price == 800],
            key=lambda c: c.id.date
        )[0]

        self.call_option_symbol = sorted(
            [c for c in contracts if c.id.option_right == OptionRight.CALL and c.id.strike_price == 600],
            key=lambda c: c.id.date
        )[0]

        self.put_option = self.add_option_contract(self.put_option_symbol)
        self.call_option = self.add_option_contract(self.call_option_symbol)

    def on_data(self, data):
        if not self.portfolio.invested and self.stock.price != 0 and self.put_option.price != 0 and self.call_option.price != 0:
            #this gets executed on start and after each auto-assignment, finally ending with expiration assignment
            if self.time < self.put_option_symbol.id.date:
                self.market_order(self.put_option_symbol, -1)

            if self.time < self.call_option_symbol.id.date:
                self.market_order(self.call_option_symbol, -1)

    def get_security(self, symbol):
        if symbol == self.stock.symbol:
            return self.stock
        if symbol == self.call_option_symbol:
            return self.call_option
        if symbol == self.put_option_symbol:
            return self.put_option
        raise RegressionTestException(f"Unexpected symbol: {symbol}")

    def on_end_of_algorithm(self):
        for trade in self.trade_builder.closed_trades:
            symbol, = trade.symbols
            direction = 1 if trade.direction == TradeDirection.LONG else -1
            multiplier = self.get_security(symbol).symbol_properties.contract_multiplier
            expected_profit_loss = round((trade.exit_price - trade.entry_price) * trade.quantity * direction * multiplier, 2)

            if trade.profit_loss != expected_profit_loss:
                raise RegressionTestException(f"Expected underlying trade profit/loss to be {expected_profit_loss}. Actual: {trade.profit_loss}")
